src/main/scala/nl.knaw.dans.easy.download/components/Statistics.scala

```diff
@@ -0,0 +1,105 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package nl.knaw.dans.easy.download.components
+
+import java.io.File
+import java.util.UUID
+
+import com.typesafe.scalalogging.Logger
+import javax.servlet.http.HttpServletRequest
+import org.slf4j.LoggerFactory
+import nl.knaw.dans.lib.logging.DebugEnhancedLogging
+
+import scala.util.Try
+import scala.xml.Elem
+
+case class Statistics(request: HttpServletRequest, bagId: UUID, fileItem: FileItem, userInfo: Option[User], ddm: Elem) extends DebugEnhancedLogging {
+
+  private val statisticsLogger = Logger(LoggerFactory.getLogger("easy-statistics"))
+  private val disciplines = EasyDisciplines.disciplines
+
+  def logDownload: Try[Unit] = Try {
+    val user = userInfo.getOrElse(User("Anonymous", Seq.empty))
+    statisticsLogger.info(getLogEventString(user))
+  }
+
+  def getLogEventString(user: User): String = {
+    val (subDiscipline, subDisciplineLabel) = getSubDiscipline
+    s"""- DOWNLOAD_FILE_REQUEST ; ${ getUserId(user) } ; roles: $getUserRoles ; groups: ${ getUserGroups(user) } ; $getIpAddress ;
+       | dataset(DATASET_ID: "$getDatasetId") ; file(FILE_NAME(0): "$getFileName") ;
+       | discipline(SUB_DISCIPLINE_ID: "${ subDiscipline }" ; TOP_DISCIPLINE_LABEL: "$getTopDisciplineLabel" ;
+       | SUB_DISCIPLINE_LABEL: "${ subDisciplineLabel }" ; TOP_DISCIPLINE_ID: "$getTopDiscipline")
+       |""".stripMargin.replace("\n", "")
+  }
+
+  private def getUserId(user: User): String = {
+    user.id
+  }
+
+  def getUserRoles: String = {
+    "(USER)"
+  }
+
+  private def getUserGroups(user: User): String = {
+    user.groups.mkString("(", ",", ")")
+  }
+
+  private def getIpAddress: String = {
+    if (request != null) {
+      val forwardedFor = request.getHeader("X-FORWARDED-FOR")
+      if (forwardedFor != null && forwardedFor.trim.nonEmpty) forwardedFor
+      else if (request.getRemoteAddr != null) request.getRemoteAddr
+      else ""
+    }
+    else ""
+  }
+
+  private def getDatasetId = {
+    val identifier = (ddm \ "dcmiMetadata" \ "identifier").find(_.text.startsWith("easy-dataset"))
+    if (identifier.isDefined) identifier.get.text
+    else datasetIdentifierNotFound
+  }
+
+  private def datasetIdentifierNotFound: String = {
+    logger.error(s"Easy dataset-id <ddm:identifier> not found in $bagId/dataset.xml")
+    "-"
+  }
+
+  private def getFileName: String = {
+    new File(fileItem.itemId).getName
+  }
+
+  private def getSubDiscipline: (String, String) = {
+    val audience = (ddm \ "profile" \ "audience").headOption
+    if (audience.isDefined) {
+      val easyDiscipline = disciplines.get(audience.get.text)
+      if (easyDiscipline.isDefined) easyDiscipline.get
+      else {
+        logger.error(s"Audience ${ audience.get.text } not found in the Map of Easy disciplines")
+        ("", "")
+      }
+    }
+    else ("", "")
```

Comment:
```scala
val discipline = for {
  audience <- (ddm \ "profile" \ "audience").headOption.map(_.text)
  easyDiscipline <- disciplines.get(audience)
} yield easyDiscipline

discipline getOrElse {
  logger.warn(s"Audience ${ audience.get.text } not found in the Map of Easy disciplines")
  ("", "")
}

Reply:
This would give an error (or warning) if there is no audience given.  Perhaps that should never occur?
In any case I have modified the code somewhat.
I'd rather give here an error if the audience is not found in the Map.  It should not happen, and if it does, we have to see what is the erason for that.

Reply:
See https://github.com/DANS-KNAW/easy-schema/blob/5879174909ae68c07a6ac683ffb5435f17894a94/lib/src/main/resources/md/ddm/ddm.xsd#L196-L206. The `audience` field is validated during schema validation. Hence you probably don't need the `disciplines.get(audience)`, resulting in an `Option[String]`. Instead you could just use `disciplines(audience)`, resulting in a `String`.

Reply:
```scala
(ddm \ "profile" \ "audience")
  .headOption
  .map(audienceNode => disciplines(audienceNode.text))
  .getOrElse { ("", "") }
```

Reply:
It is well possible that sometimes in the future a new audience is added to the classification.  The new classification in dataset.xml would cause an error if it hasn't yet been added to the static EasyDisciplines Map.
If there is always an audience in the dataset.xml, we could write:

```
  private def getSubDiscipline: (String, String) = {
    val audience = (ddm \ "profile" \ "audience").headOption.get.text
    disciplines.getOrElse(audience, audienceNotFound(audience))

  }

  private def audienceNotFound(audience: String): (String, String) = {
    logger.error(s"Audience $audience not found in the Map of Easy disciplines")
    ("", "")
  }

```

Or is `(ddm \ "profile" \ "audience").headOption.get.text` too scary?